src/contractManagers/index.ts

```diff
@@ -0,0 +1,146 @@
+import cpkFactoryAbi from '../abis/CpkFactoryAbi.json'
+import multiSendAbi from '../abis/MultiSendAbi.json'
+import safeAbiV111 from '../abis/SafeAbiV1-1-1.json'
+import safeAbiV120 from '../abis/SafeAbiV1-2-0.json'
+import { NormalizedNetworkConfigEntry } from '../config/networks'
+import EthLibAdapter, { Contract } from '../ethLibAdapters/EthLibAdapter'
+import { Address } from '../utils/basicTypes'
+import CommonContractManager from './CommonContractManager'
+import ContractV111Manager from './ContractV111Manager'
+import ContractV120Manager from './ContractV120Manager'
+
+export interface ContractManagerProps {
+  ethLibAdapter: EthLibAdapter
+  network: NormalizedNetworkConfigEntry
+  ownerAccount: Address | undefined
+  saltNonce: string
+  isSafeApp: boolean
+  isConnectedToSafe: boolean
+}
+
+class ContractManager {
+  #contractVersionManager?: CommonContractManager
+  #contract?: Contract
+  #proxyFactory: Contract
+  #masterCopyAddress: Address
+  #multiSend: Contract
+  #fallbackHandlerAddress: Address
+
+  static async create(opts: ContractManagerProps): Promise<ContractManager> {
+    const contractManager = new ContractManager(opts.ethLibAdapter, opts.network)
+    await contractManager.init(opts)
+    return contractManager
+  }
+
+  constructor(ethLibAdapter: EthLibAdapter, network: NormalizedNetworkConfigEntry) {
+    this.#masterCopyAddress = network.masterCopyAddressVersions[0].address
+    this.#fallbackHandlerAddress = network.fallbackHandlerAddress
+    this.#multiSend = ethLibAdapter.getContract(multiSendAbi, network.multiSendAddress)
+    this.#proxyFactory = ethLibAdapter.getContract(cpkFactoryAbi, network.proxyFactoryAddress)
+  }
+
+  async init(opts: ContractManagerProps): Promise<void> {
+    await this.calculateContractVersionManager(opts)
+  }
+
+  private async calculateContractVersionManager(opts: ContractManagerProps) {
+    const { ethLibAdapter, ownerAccount, saltNonce, network, isSafeApp, isConnectedToSafe } = opts
+    let proxyAddress
+    let properVersion
+
+    const salt = ethLibAdapter.keccak256(
+      ethLibAdapter.abiEncode(['address', 'uint256'], [ownerAccount, saltNonce])
+    )
+
+    if (isSafeApp || isConnectedToSafe) {
+      const temporaryContract = ethLibAdapter.getContract(safeAbiV120, ownerAccount)
+      const version: string = await temporaryContract.call('VERSION', [])
+      properVersion = version
+      proxyAddress = ownerAccount
+    } else {
+      for (const masterCopyVersion of network.masterCopyAddressVersions) {
+        proxyAddress = await this.calculateProxyAddress(
+          masterCopyVersion.address,
+          salt,
+          ethLibAdapter
+        )
+
+        const codeAtAddress = await ethLibAdapter.getCode(proxyAddress)
+        const isDeployed = codeAtAddress !== '0x'
+        if (isDeployed) {
+          properVersion = masterCopyVersion.version
```

Comment:
Maybe it would be better to set `properVersion = await temporaryContract.call('VERSION', [])` here?

Reply:
I would like to talk about this in Zoom :)